src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Subscriptions/DependenciesChanges.cs

```diff
@@ -5,32 +5,49 @@
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.Subscriptions
 {
-    internal class DependenciesChanges : IDependenciesChanges
+    internal sealed class DependenciesChanges : IDependenciesChanges
     {
-        private HashSet<IDependencyModel> Added { get; } = new HashSet<IDependencyModel>();
-        private HashSet<IDependencyModel> Removed { get; } = new HashSet<IDependencyModel>();
+        private readonly HashSet<IDependencyModel> _added = new HashSet<IDependencyModel>();
+        private readonly HashSet<IDependencyModel> _removed = new HashSet<IDependencyModel>();
 
-        public IImmutableList<IDependencyModel> AddedNodes => ImmutableArray.CreateRange(Added);
-        public IImmutableList<IDependencyModel> RemovedNodes => ImmutableArray.CreateRange(Removed);
+        public bool AnyChanges => _added.Count != 0 || _removed.Count != 0;
 
-        public void IncludeAddedChange(IDependencyModel model)
+        public ImmutableArray<IDependencyModel> AddedNodes
         {
-            Added.Add(model);
+            get
+            {
+                lock (_added)
+                {
+                    return ImmutableArray.CreateRange(_added);
```

Comment:
Is there benefit to move these to immutable to avoid the copying, and then use ApplyChangeOptimistically to avoid the locking?

Reply:
Possibly. I did consider it. I felt this approach might come out ahead, but I should do some measurements to find out. With the optimisic pattern we'd need to use `ImmutableList<T>` rather than `ImmutableArray<T>`, which has worse space complexity.

I experimented with separating these two phases out (builder and snapshot), but `DependenciesRuleChangeContext` _can_ be modified after would first need the snapshot. There's possibly something more we can do here.

Reply:
I ran a benchmark to learn more about this.

Here's the code:

```c#
[ClrJob]
[MemoryDiagnoser]
public class BuildImmutableFromSetBenchmarks
{
    [Params(0, 1, 10, 100, 1000)]
    public int Count;

    private string[] _strings;
    private HashSet<string> _hashSet;
    private ImmutableHashSet<string> _immutableHashSet;

    [GlobalSetup]
    public void Setup()
    {
        _strings = Enumerable
            .Range(0, Count)
            .Select(i => i.ToString())
            .ToArray();
    }

    [Benchmark]
    public ImmutableHashSet<string> OptimisticImmutableHashSet()
    {
        _immutableHashSet = ImmutableHashSet.Create<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var s in _strings)
        {
            var before = Volatile.Read(ref _immutableHashSet);
            while (Interlocked.CompareExchange(ref _immutableHashSet, before.Add(s), before) != before)
            {

            }
        }

        return _immutableHashSet;
    }

    [Benchmark]
    public ImmutableArray<string> HashSetToImmutableArray()
    {
        _hashSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var s in _strings)
        {
            lock (_hashSet)
            {
                _hashSet.Add(s);
            }
        }

        lock (_hashSet)
        {
            return ImmutableArray.CreateRange(_hashSet);
        }
    }
}
```

I tried to replicate the scenario as best as possible. Obviously there's no concurrency here. It also doesn't simulate producing the `ImmutableArray` more than once, which can happen in theory. Actually we can cache the result, with invalidation on write to help with that.

The results:

|                           Method | Count |           Mean |         Error |        StdDev |         Median |   Gen 0 |  Gen 1 | Allocated |
|--------------------------------- |------ |---------------:|--------------:|--------------:|---------------:|--------:|-------:|----------:|
|       OptimisticImmutableHashSet |     0 |       621.5 ns |     12.103 ns |     15.306 ns |       616.6 ns |  0.0029 |      - |      20 B |
|          HashSetToImmutableArray |     0 |       145.2 ns |      3.935 ns |     11.603 ns |       138.6 ns |  0.0074 |      - |      40 B |
|       OptimisticImmutableHashSet |     1 |       848.4 ns |     16.019 ns |     20.829 ns |       849.9 ns |  0.0134 |      - |      72 B |
|          HashSetToImmutableArray |     1 |       275.4 ns |      5.205 ns |      5.112 ns |       276.9 ns |  0.0286 |      - |     152 B |
|       OptimisticImmutableHashSet |    10 |     4,547.7 ns |     89.233 ns |    158.612 ns |     4,555.7 ns |  0.2289 |      - |    1212 B |
|          HashSetToImmutableArray |    10 |     1,579.6 ns |     19.275 ns |     17.086 ns |     1,583.4 ns |  0.1144 |      - |     620 B |
|       OptimisticImmutableHashSet |   100 |    71,160.5 ns |  1,463.607 ns |  4,055.647 ns |    70,170.6 ns |  4.3945 |      - |   23300 B |
|          HashSetToImmutableArray |   100 |    13,689.1 ns |    264.648 ns |    325.012 ns |    13,730.5 ns |  1.1749 |      - |    6220 B |
|       OptimisticImmutableHashSet |  1000 | 1,035,607.0 ns | 17,529.009 ns | 14,637.515 ns | 1,032,324.7 ns | 68.3594 | 3.9063 |  360931 B |
|          HashSetToImmutableArray |  1000 |   131,551.6 ns |  2,597.237 ns |  2,027.752 ns |   130,749.8 ns | 11.7188 | 1.2207 |   62429 B |

```text
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.345 (1803/April2018Update/Redstone4)
Intel Xeon Silver 4110 CPU 2.10GHz, 2 CPU, 32 logical and 16 physical cores
Frequency=2045972 Hz, Resolution=488.7652 ns, Timer=TSC
  [Host] : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3190.0
  Clr    : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3190.0

Job=Clr  Runtime=Clr
```

Summary from those results is that the `HashSet` approach, with `Monitor`-based locking, is strictly faster on all counts, however for lower counts it allocates more memory. As the number of items grows however, the optimistic immutable collection approach allocates considerably more. Somewhere between 1 and 10 elements, the `HashSet` approach takes the lead.

![image](https://user-images.githubusercontent.com/350947/47793610-98f4d000-dd16-11e8-8528-18d524647dec.png)

Reply:
What kind of contention do we expect to see here? It's not clear me to how many threads are actually looking that are adding nodes and inspecting them? I guess not a lot for the latter given you're not worried about the ImmutableArray allocations?